Clarify cleanup in docs. (#13381)
ethereum/solidity: docs/internals/variable_cleanup.rst

|signed integers|higher bits    |currently silently       |
|               |set to the     |signextends to a valid   |
|               |sign bit       |value, i.e. all higher   |
|               |               |bits are set to the sign |
|               |               |bit; may throw an        |
|               |               |exception in the future  |
+---------------+---------------+-------------------------+
|unsigned       |higher bits    |currently silently masks |
|integers       |zeroed         |to a valid value, i.e.   |
|               |               |all higher bits are set  |
|               |               |to zero; may throw an    |
|               |               |exception in the future  |
+---------------+---------------+-------------------------+

Note that valid and invalid values are dependent on their type size.
Consider ``uint8``, the unsigned 8-bit type, which has the following valid values:

.. code-block:: none

    0000...0000 0000 0000
    0000...0000 0000 0001
    0000...0000 0000 0010
    ....
    0000...0000 1111 1111

Any invalid value will have the higher bits set to zero:

.. code-block:: none

    0101...1101 0000 0000   invalid value

>     0000...0000 0000 0000   cleaned value

Reviewer: Maybe add signed integer examples to clarify the "signextend" term?

```suggestion
    0000...0000 0000 0000   cleaned value
    
For positive 8-bit signed integers:
.. code-block:: none
    Valid values:
    0000...0000 0000 0001
    0000...0000 0000 0010
    ....
    0000...0000 0111 1111

    invalid value: 0101...1101 0000 0100
    sign bit: 0
    cleaned value: 0000...0000 0000 0100

For negative 8-bit signed integers:
.. code-block:: none
    Valid values:
    1111...1111 1111 1111
    1111...1111 1111 1110
    ....
    1111...1111 1000 0000

    invalid value: 0000...0000 1111 1111
    sign bit: 1
    cleaned value: 1111...1111 1111 1111
```

Reviewer 2: Done.

> Different types have different rules for cleaning up invalid values:

Reviewer 2: ```suggestion
The following table describes the rules applied for valid and invalid values of different types, where ``higher bits`` refers to the remaining bits in case the type has less than 256 bits.
```

+---------------+---------------+-------------------------+
|Type           |Valid Values   |Cleanup of Invalid Values|
+===============+===============+=========================+
|enum of n      |0 until n - 1  |throws exception         |
|members        |               |                         |
+---------------+---------------+-------------------------+
|bool           |0 or 1         |results in 1             |
+---------------+---------------+-------------------------+

> |signed integers|higher bits    |currently silently       |

Reviewer 3: Arent' valid values for signed integers `(-2**(n-1), 2**(n-1)]?

Reviewer 2: I think he meant that valid values have their `higher bits = 256 - n` set to the sign bit... which is the same for invalid values...

Reviewer 4: Marking as resolved.

|               |word           |signextends to a valid   |
|               |               |value, i.e. all higher   |
|               |               |bits are set to the sign |
|               |               |bit; may throw an        |
|               |               |exception in the future  |
+---------------+---------------+-------------------------+
|unsigned       |higher bits    |currently silently masks |

> |integers       |zeroed         |to to a valid value, i.e.|

Reviewer: ```suggestion
|integers       |zeroed         |to a valid value, i.e.|
```

> |               |               |all higher bits are set  |

Reviewer 3: What does `higher bits zeroed` mean for unsigned integers wrt valid values? Aren't valid values [0, 2**n - 1]?

Reviewer 2: The table is describing the rule applied if the value is valid, i.e, fits the `n` bits length of the type. Currently, the same is done for invalid values also.
So, in cases where the type has less than 256 bits, the `higher = 256 - n` bits will be zeroed.

Reviewer 4: Marking as resolved.
